ComponentKit/StatefulViews/CKStatefulViewReusePool.mm

```diff
@@ -16,46 +16,34 @@
 class FBStatefulReusePoolItem {
 public:
   FBStatefulReusePoolItem()
-  : preferredSuperviewMap([NSMapTable weakToStrongObjectsMapTable]), allViews([NSMutableArray array]) {};
+  : views([NSMutableArray array]) {};
 
   UIView *viewWithPreferredSuperview(UIView *preferredSuperview)
   {
-    NSMutableArray *matchingViews = [preferredSuperviewMap objectForKey:preferredSuperview];
-    UIView *view = [matchingViews lastObject] ?: [allViews lastObject];
-    if (view) {
-      if ([matchingViews count]) {
-        [matchingViews removeObject:view];
-      } else {
-        [[preferredSuperviewMap objectForKey:[view superview]] removeObject:view];
-      }
-      [allViews removeObject:view];
+    if ([views count] == 0) {
+      return nil;
     }
+    const NSUInteger matchingIndex = [views indexOfObjectPassingTest:^BOOL(UIView *view, NSUInteger idx, BOOL *stop) {
+      return [view superview] == preferredSuperview;
+    }];
+    const NSUInteger viewIndex = (matchingIndex == NSNotFound) ? 0 : matchingIndex;
```

Comment:
In the previous implementation it looks like view selection would fall back onto the last view in `allViews`. Do we want to fall back onto the first view in `views` here?